### PR TITLE
Ansible kind automation

### PR DESCRIPTION
**Pull request by ssharmasandeep**
Ansible playbook to create kind clusters and deploy required CNIs

### CODE REVIEW

File: ansible_kind/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+- This ansible should be triggered from the kind-cluster-gke terraform plan. So that we have one command to create the Infra.
+
+
+## Requirements
+- [ansible](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html#pip-install)
+
+## Prerequisites
+- Terraform plan executed already.
+- External IP address of the GCP VM.
+- ssh key for loging in to the VM. FIXME I am using the the key called google_compute_engine which is getting created after I execute "gcloud compute ssh --zone "us-central1-a" "nephio-poc-001"  --project "pure-faculty-367518"gcloud compute ssh --zone "us-central1-a" "nephio-poc-001"  --project "pure-faculty-367518""
```
Comment (debeaueric) [inline ansible_kind/README.md]:
```suggestion
- ssh key for loging in to the VM. FIXME I am using the the key called google_compute_engine which is getting created after I execute "gcloud compute ssh --zone "us-central1-a" "nephio-poc-001"  --project "pure-faculty-367518"
```

File: ansible_kind/kind_setup.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+     copy: 
+      src: ./scripts/multiclusterkind 
+      dest: /home/ubuntu/ 
+      owner: ubuntu
+      group: ubuntu
+      mode: 0644
+   - name: Prepare VM for kind installation
+     command: bash /home/ubuntu/multiclusterkind/k8s_cluster_setup.sh    
+   - name: Install kind clusters 
+     command: bash /home/ubuntu/multiclusterkind/kind_create_clusters.sh 2
```
Comment (debeaueric) [inline ansible_kind/kind_setup.yaml:19]: Only 2 clusters ?

File: ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+cat << EOF > "$FILE"
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+networking:
+  podSubnet: "{{pod_subnet}}"
+  serviceSubnet: "{{service_subnet}}"
+featureGates:
+  EndpointSliceProxying: true
+nodes:
+  - role: control-plane
```
Comment (debeaueric) [inline ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh:64]: No worker ?

**Review by debeaueric — CHANGES_REQUESTED** [review]
Some comments on the number of clusters created and workers

File: ansible_kind/kind_setup.yaml (hunk 2 above)
Comment (ssharmasandeep) [reply ansible_kind/kind_setup.yaml:19]: We can update it to the number we decide collectively, for testing I set it to 2.

File: ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh (hunk 3 above)
Comment (ssharmasandeep) [reply ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh:64]: If we need clusters with worker nodes, then I can update this.

File: ansible_kind/scripts/multiclusterkind/k8s_cluster_setup.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    local USER="$1"	
+    sudo apt-get -y update
+    sudo apt-get -y install docker.io apt-transport-https make gcc cmake
+    sudo systemctl enable docker.service    
+    sudo usermod -aG docker "${USER}"
+    sudo chmod 777 /var/run/docker.sock
+}
+
+function install_kind() {
+   curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.11.1/kind-linux-amd64
```
Comment (vjayaramrh) [inline ansible_kind/scripts/multiclusterkind/k8s_cluster_setup.sh]: @sandeepAarna FYI in the call yesterday, it was discussed to use the 0.17 version of kind.
Is your testing done only with 0.11? Do you have any concerns updating to 0.17?

File: ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+NUM_CLUSTERS=${1:-4}
+CLUSTER_VERSION=${CLUSTER_VERSION:-"kindest/node:v1.21.1"}
+KIND_LOG_FILE=${KIND_LOG_FILE:-"/tmp/targetClusters"}
+
+# Make sure go exists and the go version is a viable version.
+util::cmd_must_exist "go"
+util::verify_go_version
+
+# install kind and kubectl
+kind_version=v0.11.1
```
Comment (vjayaramrh) [inline ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh:28]: same comment about updating to 0.17 as above

File: ansible_kind/scripts/multiclusterkind/k8s_cluster_setup.sh (hunk 4 above)
Comment (ssharmasandeep) [reply ansible_kind/scripts/multiclusterkind/k8s_cluster_setup.sh]: what is the dependency on .17 ?

Comment (vishwanathj) [reply ansible_kind/scripts/multiclusterkind/k8s_cluster_setup.sh]: > what is the dependency on .17 ?

@electrocucaracha @debeaueric @johnbelamaric Are you aware of any dependencies on 0.17? @sandeepAarna scripts install the 0.11.1 KIND version currently

**Review by electrocucaracha — COMMENTED** [review]
@sandeepAarna what is the difference between these scripts and the instructions placed [here](https://github.com/nephio-project/one-summit-22-workshop/blob/main/scripts/install.sh)? It seems like both install docker, kind, kubectl and deploy KinD clusters.

File: ansible_kind/scripts/multiclusterkind/util.sh [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+  local container_name=$1
+  docker inspect --format='{{range $key, $value := index .NetworkSettings.Ports "6443/tcp"}}{{if eq $key 0}}{{$value.HostIp}}:{{$value.HostPort}}{{end}}{{end}}' "${container_name}"
+}
+
+# util::check_clusters_ready checks if a cluster is ready, if not, wait until timeout
+function util::check_clusters_ready() {
+  local kubeconfig_path=${1}
+  local context_name=${2}
+
+  echo "Waiting for kubeconfig file ${kubeconfig_path} and clusters ${context_name} to be ready..."
```
Comment (vjayaramrh) [inline ansible_kind/scripts/multiclusterkind/util.sh]: @sandeepAarna I tried just the standalone bash scripts in the VM (vishsandbox) on GCP cluster and noted that the fourth cluster is not getting created
Screen Shot 2022-11-08 at 4 53 53 PM

File: ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+do
+  FILE="${TEMP_PATH}"/cluster${i}.yaml	
+cat << EOF > "$FILE"
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+networking:
+  podSubnet: "{{pod_subnet}}"
+  serviceSubnet: "{{service_subnet}}"
+featureGates:
+  EndpointSliceProxying: true
```
Comment (electrocucaracha) [inline ansible_kind/scripts/multiclusterkind/kind_create_clusters.sh]: This feature gate came GA in [1.22](https://v1-22.docs.kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
```suggestion
```

File: ansible_kind/kind_setup.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  gather_facts: no
+  strategy: free
+  user: ubuntu
+  tasks:
+   - name:
+     ansible.builtin.wait_for_connection:
+   - name: Copy multuclusterkind directory
+     copy: 
+      src: ./scripts/multiclusterkind 
+      dest: /home/ubuntu/
```
Comment (vishwanathj) [inline ansible_kind/kind_setup.yaml:12]: Is the assumption that a user named `ubuntu` will be created by terraform scripts? OR would this need to be a parameter passed?

Comment (ssharmasandeep) [reply ansible_kind/kind_setup.yaml:12]: It is a ubuntu VM, so a ubuntu user will be there.. if we want to create more users then either the terraform needs to enhanced .

**Pull request merged**